common-kafka/src/test/java/com/cerner/common/kafka/consumer/ProcessingKafkaConsumerRebalanceIT.java

```diff
@@ -0,0 +1,600 @@
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Integration test for a specific deadlocking condition in ProcessingKafkaConsumer client after repeated rebalances.
+ */
+public class ProcessingKafkaConsumerRebalanceIT {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ProcessingKafkaConsumerITest.class);
+
+    private static final long MAX_SLEEP = 150; // 150 ms
+    private static final long CONSUMER_POLL_TIMEOUT = 1000; // 1s
+    private static final long CONSUMER_THREAD_JOIN_TIMEOUT = 5000; // 5s - max time we wait for consumer thread to finish running
+    private static final int PARTITIONS = 1;
+    private static final int MESSAGES_PER_TOPIC = 25;
+    private static final long HISTORY_CHECK_TIME = 1000; // 1s - how often to check/print processing history during testing
+
+    private static Properties CONSUMER_PROPERTIES = new Properties();
+
+    private static KafkaAdminClient kafkaAdminClient;
+
+    @Rule
+    public TestName name = new TestName();
+
+    @BeforeClass
+    public static void startup() throws Exception {
+        KafkaTests.startTest();
+        kafkaAdminClient = new KafkaAdminClient(KafkaTests.getProps());
+
+        CONSUMER_PROPERTIES.putAll(KafkaTests.getProps());
+        CONSUMER_PROPERTIES.setProperty(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG, "3000");
+        CONSUMER_PROPERTIES.setProperty(ConsumerConfig.MAX_POLL_RECORDS_CONFIG, "1");
+        CONSUMER_PROPERTIES.setProperty(ConsumerConfig.PARTITION_ASSIGNMENT_STRATEGY_CONFIG, FairAssignor.class.getName());
+        CONSUMER_PROPERTIES.setProperty(ProcessingConfig.COMMIT_SIZE_THRESHOLD_PROPERTY, "1"); // >= 1 offsets in a partition
+        CONSUMER_PROPERTIES.setProperty(ProcessingConfig.COMMIT_TIME_THRESHOLD_PROPERTY, "1000"); // >= 1s since last commit
+        CONSUMER_PROPERTIES.setProperty(ProcessingConfig.FAIL_PAUSE_TIME_PROPERTY, "100"); // 100 ms
+        CONSUMER_PROPERTIES.setProperty(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+        CONSUMER_PROPERTIES.setProperty(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+        CONSUMER_PROPERTIES.remove(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG);
+        CONSUMER_PROPERTIES.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+    }
+
+    @AfterClass
+    public static void shutdown() throws Exception {
+        kafkaAdminClient.close();
+        KafkaTests.endTest();
+    }
+
+
+    @Test
```

Comment:
Nit: Add a timeout attribute to the test since it looks like we have a while loop that doesn't quit unless we see all our data